complete rewrite of niceTables.pl

Alex-Jordan opened the pull request:
This is a complete rewrite of niceTables.pl.

* I'm not sure how to write POD. If this is not good, can anyone point me to a reference?
* I've been out of development for a while. Do I need to run perltidy on this?
* This should only change the behavior from the previous version in small ways that no one should care about. A few new features are added.
* Here is an example .pg file that minimally runs through features. (Well, I had to change its extension to .txt for GitHub to let me upload it.) Testing should examine both HTML and hardcopy output (and PTX if you want to). Should I put this in the testing folder as part of this PR?


[newNiceTables.txt](https://github.com/openwebwork/pg/files/10193082/newNiceTables.txt)

pstaabp commented on macros/niceTables.pl:
```diff
@@ -1,837 +1,987 @@
+# encase, rowheaders should be passed to cells
+# various css should be self-explanatory
+sub TableEnvironment {
+	my $dataArray = shift;
```
We're trying to switch to:
```
my ($dataArray, $optsArray, $colCount, $tableOpts, $alignment) = @_; 
```
to pull arguments from the function calls.  It's shorter and this way this ensures there are no other arguments later in the function.

pstaabp commented on macros/niceTables.pl:
```diff
@@ -1,837 +1,987 @@
+		bottom => 0,
+		valign => '',
+	);
+	my $originalArrayRef = shift;
```
Put this line at the top of the function.

pstaabp commented:
You will need to run perltidy on this.

I'd like to see the pg file you attached turned into a unit test.  I think we don't have examples with html output in the tests yet, but it would be good to develop such tests.  I can help look into this.

Alex-Jordan commented:
I changed how arguments are handled, and I used pertidy. (I assume I should use it with all default options?)

But I still see github marking the perltidy test as failed...

drgrice1 commented:
The `.perltidyrc` file in the pg root directory needs to be used. You can just run perltidy with the `bin/run-perltidy` script and it will take care of the settings.

Also note that you need to make sure that you have a new enough version of perltidy installed.  I believe the version in the Ubuntu repositories is not new enough (even on Ubuntu 22.04).  You need at least version 20220613.  So you will need to install that from cpan.

Alex-Jordan commented:
I'm on a Mac and used `brew install perltidy` and the version I got was v20221112. So that part is OK. After running with with `bin/run-perltidy`, it passes the GitHub check now.

pstaabp commented:
@Alex-Jordan One of the recent PRs moved all of the macros into subdirectories, so you'll need to adjust this to that structure.  Also, I put in a PR to your branch with a sample test file that had a few bug fixes as well.

drgrice1 commented:
I am seeing a lot of warnings  with this right now.  You probably are seeing these too though.

The attached problem has a basic table.  I note a difference in the resulting displayed table from the develop branch version of niceTables.pl.  That is a border on the bottom of the `thead` section that wasn't there before.
[niceTablesTest.pg.txt](https://github.com/openwebwork/pg/files/10352598/niceTablesTest.pg.txt)

On POD, I recommend removing all of the `#` characters that seem to be used as a separator.  The alignment looks good on its own, and those characters just add noise.

Also, I recommend not using any tab characters in POD.  Tab characters are taken verbatim in POD and are assumed to be on 8-column boundaries.  That usually doesn't end up looking very good in the generated output from POD formatting programs.

Please remove the multiple empty lines that are in the POD.  A single empty line before and after any POD directive (lines starting with `=`) is required.  Anything more than makes POD take up more space than it needs to in the code.  POD formatting utilities remove all of those lines from the generated POD in any case.  perltidy removes multiple empty lines outside of POD, but perltidy won't touch anything inside POD.

Useful utilities for POD are `podchecker`  and `perldoc` which are included in most perl distributions by default.  You can also use the `bin/dev_scripts/generate-ww-pg-pod.pl` script from webwork2.  Example usage:
```bash
./bin/dev_scripts/generate-ww-pg-pod.pl -o ~/pod
```

Comment moved here!  Sorry.

drgrice1 commented:
There are a couple other changes that I have noticed in testing this.

The first is demonstrated in the attached file.  
[synthetic-division.pg.txt](https://github.com/openwebwork/pg/files/10504043/synthetic-division.pg.txt)
This uses niceTables.pl to display synthetic division.  With the version of niceTables.pl in develop, the vertical line in the `halign` attribute of the second column appears to the left of that column, but with the version in this pull request it does not.

The other change I support, if that change was intentional.  I noticed that with this pull request the `\toprule` and `\bottomrule` that were added in TeX for hardcopy are gone.  I never like those, and in fact I used a regular expression substitution to remove those.

Alex-Jordan commented:
First, I'm sorry to be slow on this PR. Acting on your and Peter's feedback
is on my radar, just keeps getting lower priority.

I want to move toward PreTeXt's philosophy for table rules, which is:
cell can have right border; bottom border. (no left or top)
row can have left border; bottom border. (no right or top)
cell can have right border; top border. (no left or bottom)
table can have left border; top border. (it can have right but that is
interpreted for convenience as applying right to each col; it can have
bottom but that is interpreted as applying bottom to each row)

This means sometimes what you think of a cell's left border is "actually"
the right border of the adjacent cell, or it is the left border of the row.
Etc. Without the clarity from this simplification, it is a mess to work out
if the line between two cells should be one cell's right border or the
other one's left border, or both. And then that matters for the PTX
translation, but also when you want a cell attribute to override something
that applied to the surrounding row, col, or table.

So if it is a problem that this changes behavior for the kind of example
you mention, then I'd want to leave this as it is but name it
niceTables2.pl or something. Of course one could edit the problem to make
that line be the previous cell's right border, but I understand we don't
want people to have to modify their problem code. Either way, I know I need
to make things clear in the documentation.

Yes, removing the forced toprule and bottomrule is intentional.



On Wed, Jan 25, 2023 at 2:03 PM Glenn Rice ***@***.***> wrote:

> There are a couple other changes that I have noticed in testing this.
>
> The first is demonstrated in the attached file.
> synthetic-division.pg.txt
> <https://github.com/openwebwork/pg/files/10504043/synthetic-division.pg.txt>
> This uses niceTables.pl to display synthetic division. With the version of
> niceTables.pl in develop, the vertical line in the halign attribute of
> the second column appears to the left of that column, but with the version
> in this pull request it does not.
>
> The other change I support, if that change was intentional. I noticed that
> with this pull request the \toprule and \bottomrule that were added in
> TeX for hardcopy are gone. I never like those, and in fact I used a regular
> expression substitution to remove those.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openwebwork/pg/pull/752#issuecomment-1404284449>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABEDOAABPMWYVPBV76W2WP3WUGPJFANCNFSM6AAAAAASZCMM2E>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Alex Jordan
Mathematics Instructor
Portland Community College

Alex-Jordan commented:
In that last post there is a copy-paste error:
*col* can have right border; top border. (no left or bottom)

On Wed, Jan 25, 2023 at 3:47 PM Alex Jordan ***@***.***> wrote:

> First, I'm sorry to be slow on this PR. Acting on your and Peter's
> feedback is on my radar, just keeps getting lower priority.
>
> I want to move toward PreTeXt's philosophy for table rules, which is:
> cell can have right border; bottom border. (no left or top)
> row can have left border; bottom border. (no right or top)
> cell can have right border; top border. (no left or bottom)
> table can have left border; top border. (it can have right but that is
> interpreted for convenience as applying right to each col; it can have
> bottom but that is interpreted as applying bottom to each row)
>
> This means sometimes what you think of a cell's left border is "actually"
> the right border of the adjacent cell, or it is the left border of the row.
> Etc. Without the clarity from this simplification, it is a mess to work out
> if the line between two cells should be one cell's right border or the
> other one's left border, or both. And then that matters for the PTX
> translation, but also when you want a cell attribute to override something
> that applied to the surrounding row, col, or table.
>
> So if it is a problem that this changes behavior for the kind of example
> you mention, then I'd want to leave this as it is but name it
> niceTables2.pl or something. Of course one could edit the problem to make
> that line be the previous cell's right border, but I understand we don't
> want people to have to modify their problem code. Either way, I know I need
> to make things clear in the documentation.
>
> Yes, removing the forced toprule and bottomrule is intentional.
>
>
>
> On Wed, Jan 25, 2023 at 2:03 PM Glenn Rice ***@***.***>
> wrote:
>
>> There are a couple other changes that I have noticed in testing this.
>>
>> The first is demonstrated in the attached file.
>> synthetic-division.pg.txt
>> <https://github.com/openwebwork/pg/files/10504043/synthetic-division.pg.txt>
>> This uses niceTables.pl to display synthetic division. With the version
>> of niceTables.pl in develop, the vertical line in the halign attribute
>> of the second column appears to the left of that column, but with the
>> version in this pull request it does not.
>>
>> The other change I support, if that change was intentional. I noticed
>> that with this pull request the \toprule and \bottomrule that were added
>> in TeX for hardcopy are gone. I never like those, and in fact I used a
>> regular expression substitution to remove those.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/openwebwork/pg/pull/752#issuecomment-1404284449>, or
>> unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ABEDOAABPMWYVPBV76W2WP3WUGPJFANCNFSM6AAAAAASZCMM2E>
>> .
>> You are receiving this because you were mentioned.Message ID:
>> ***@***.***>
>>
>
>
> --
> Alex Jordan
> Mathematics Instructor
> Portland Community College
>


-- 
Alex Jordan
Mathematics Instructor
Portland Community College

drgrice1 commented:
I don't mind modifying my problems if need be.  I have 68 problems that use niceTables.pl, and probably not all are affected by this issue.

However, can you provide the PreTeXt rationale for this philosophy?  Furthermore, why should that philosophy apply to webwork, and is it important enough to justify breaking compatibility?

I definitely do not want to see a niceTables2.pl.  There is enough of a mess caused by creating versioned file names already.  For example, all of the compoundProblem?.pl macros need to be removed, and any problem that uses them updated to use scaffold.pl.  I have finally eliminated all of the alternate versions of SetMaker.pm, ProblemSetDetail.pm, and et al. in the webwork2 code.  I don't want to see that happen again.  Fix the existing version in such a way that all are satisfied, and never make another versioned file name.  That just results in a maintenance nightmare.

Alex-Jordan commented:
Sure, there are about a half dozen small reasons, and if I were clever enough, they probably all have some umbrella explanation. But for something specific, consider:

```
\begin{tabular}{c!{\vrule width 1ex}c!{\vrule width 1ex}c}
a&\multicolumn{1}{!{\vrule width 2ex}c!{\vrule width 2ex}}{b}&c\\
d&e&f
\end{tabular}
```

which comes out like:
<img width="104" alt="Screen Shot 2023-01-25 at 6 55 49 PM" src="https://user-images.githubusercontent.com/4732672/214748455-3b0ae966-99f2-42c6-aa52-fea777c2f9f1.png">

You would think based on the coding that the middle cell in the top row has a left border that is 2ex wide, and a right border that is 2ex wide. Now how do these interact with the overall column borders that are 1ex wide? Well we can see that with that cell's right border, it gets melded in with the  overall column border and you really do have a 2ex right border. But on the left, they are adjacent and the effect is a 3ex border.

Just not letting cells have left borders in the first place gets around this.

There are a handful of other things.  With general output forms, asking questions of precedence (does a column attribute trump a row attribute or the other way round?) and questions about merging versus butting up against each other like this tex example make messy situations and it's all resolved by the outline that PreTeXt has for this. (All credit to Rob for reasoning it out.)

Alex-Jordan commented:
In that example, it would be more correct to say the cell's right border overrides the column border, but the left border won't override it. A more practical example would be like when you want borders between all columns, but not in the top row. In this example:

```
\begin{tabular}{c|c|c}
\multicolumn{1}{c}{a}  &  b  &\multicolumn{1}{c}{c}\\
d&e&f\\
g&h&i
\end{tabular}
```

The first `multicolumn` succeeds at overriding the vertical rule between columns 1 and 2. But the second `multicolumn` fails to override the vertical rule between columns 2 and 3. And with:
 
```
\begin{tabular}{c|c|c}
a  &  \multicolumn{1}{c}{b}  & c\\
d&e&f\\
g&h&i
\end{tabular}
```

it is the opposite situation.

drgrice1 reviewed (commented):
One change that I would like to see made to this macro is to make this a package.  The reason this is needed is that you have added several methods that are intended for internal use.  By making this a package that would scope those methods.  Then you would need to "export" the methods that are intended to be used externally.  This would be done by adding lines like
```
main::PG_restricted_eval('sub DataTable { NiceTables::DataTable(@_) }');
```
to the init method for all methods you want to be available externally.  This assumes you name the package `NiceTables`.  This package would not need a `new` method since it doesn't need an object.

The point here is that those added "internal" methods may cause conflicts with same named methods in other macros.  Not that there may not be any at this point, but it is likely there may be in the future.  Particularly for the simply named `tag` and `css` methods. In fact, I was thinking that a `tag` method would be nice for `PGbasicmacros.pl` to make the methods in that macro more versatile and the method would be useful elsewhere as well.  Your method has some things that are clearly specially for this macro (

Alex-Jordan commented:
I ran through the POD here to make it better. More changes coming.

Alex-Jordan commented:
I've never written an object oriented package in perl, and I've only done it in python while taking a programming course (so not in "real life"). I attempted to do this here, and the file is still working to render the problem correctly. But I have little confidence that I did this correctly. Do you have any feedback about how it is structured now with `package NiceTables`?

I know I need perltidy, and apparently some unit test failed. And I need to get Peter's new unit test.

Alex-Jordan commented:
Of course I am just now seeing that you gave some guidance in your last post.

drgrice1 commented:
In this case you don't need an object, just a name space.  So there is no need for object oriented thinking.  I was suggesting that you declare the NiceTables package, and add the following lines to `_niceTables_init` to get
```
sub _niceTables_init {
	main::PG_restricted_eval('sub DataTable { NiceTables::DataTable(@_) }');
	main::PG_restricted_eval('sub LayoutTable { NiceTables::LayoutTable(@_ }');
}
```
However, your approach should work equally well.   In fact I like it better.  `PG_restricted_eval` is really not needed here.  I would like to see it used less than it is anyway.

drgrice1 commented:
By the way, the POD looks much better!

drgrice1 commented on macros/ui/niceTables.pl:
```diff
@@ -3,835 +3,1201 @@ =head1 niceTables.pl
+		if ($token =~ /^\|/) {
+
+			# this counts how many | we have
+			$align[0]->{left} = 0
+				unless ($align[$i]->{left} && $align[$i]->{left} =~ /\d+/);
```
I believe this should be
```suggestion
			$align[0]{left} = 0
				unless ($align[0]{left} && $align[0]{left} =~ /\d+/);
```
Note:  Once you access an array or hash by the first index or key, there is no need for further dereferencing via `->`.  It doesn't do anything but make the line longer.  An array or hash can not contain non-scalar values that don't need to be dereferenced, so it doesn't add clarity to the code either.

drgrice1 commented on macros/ui/niceTables.pl:
```diff
@@ -3,835 +3,1201 @@ =head1 niceTables.pl
+	$separator = "\n" unless defined $separator;
+	my $return = "<$name";
+	for my $x (main::lex_sort(keys %$attributes)) {
+		$return .= qq( $x="$attributes->{$x}") if ($attributes->{$x} ne '');
```
`$attributes->{x}` can be undefined and gives more warnings about the use of an unitialized value in string comparison.

drgrice1 commented on macros/ui/niceTables.pl:
```diff
@@ -3,835 +3,1201 @@ =head1 niceTables.pl
+sub getPTXthickness {
+	my $input  = shift;
+	my $output = '';
+	if ($input == 1) {
+		$output = "minor";
+	} elsif ($input == 2) {
+		$output = "medium";
+	} elsif ($input >= 3) {
+		$output = "major";
+	} elsif ($input eq '0.04em') {
+		$output = 'minor';
+	} elsif ($input eq '0.07em') {
+		$output = 'medium';
+	} elsif ($input eq '0.11em') {
+		$output = 'major';
```
In all of these conditionals `$input` can be undefined, and if so gives warnings about the use of an uninitialized value in string or numeric comparison.

drgrice1 commented on macros/ui/niceTables.pl:
```diff
@@ -3,835 +3,1201 @@ =head1 niceTables.pl
+		# determine if this column has paragraph cells
+		my $width = '';
+		for my $y (@$optsArray) {
+			if ($y->[$i]->{halign} =~ /^p\{([^}]*?)\}/) {
```
`$y->[$i]->{halign}` can be undefined, and if so gives warnings about the use of an uninitialized value in pattern match.

drgrice1 reviewed (commented):
This is a list of all of the places that I am seeing that are causing warnings.  I doubt that this list is exhaustive.  These are just the warnings that I have seen in a few cases.  Different usage may bring more warnings in other places as well.

drgrice1 commented:
By the way, the unit test that is failing is caused by this macro emitting warnings when it is loaded.  The warnings are those regarding the unescaped left braces in the regular expressions.

Alex-Jordan commented:
Iterating closer to completion. I redesigned some parts of the code to really make sure that everything is initialized to be rid of most of the perl warnings. I had thought it was already designed that way, but I had overlooked a few things.

I added the `PG_restricted_eval` lines as you suggested.

I added a few features to make a few things easier and/or use better keywords. You can see these in the top lines of the diff for c1d0a36.

I need to add the unit test. And I would like to read through my code comments and revise them. But this is otherwise ready for some testing and critique.

Alex-Jordan commented:
Here is my current testing file. [niceTables.txt](https://github.com/openwebwork/pg/files/10834836/niceTables.txt) I'm looking into making it into a unit test like @pstaabp  did with an earlier version.

drgrice1 commented:
This is looking much better now.  The code is so much cleaner than it was before, and the POD much improved.

Alex-Jordan commented:
I am doing a thorough examination of the PTX output from the comprehensive test problem and I see some issues to patch. Should be independent from HTML and TeX.

pstaabp commented:
@Alex-Jordan About the testing.  I'm not sure it's needed so much for this due to the visual nature of this package.  I originally caught a lot of errors in just loading the package and creating tables, but these are all cleaned up now.  

I also put in a PR to your branch that does some other formatting for the POD which I think helps the readability of the POD.

drgrice1 commented on t/output_macros/niceTables.pg:
```diff
@@ -0,0 +1,452 @@
+$path = insertGraph($drawing);
+
+
+$lorem = "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.";
```
This line is too long.  No line should exceed 120 characters.  In addition there is excessive white space between `C<DataTable()>` and the text after it.  There really only needs to be one space there.  Generally in POD I would like to see even shorter than 120 characters.  This is really the same as what you have said about problem text being to wide.  It detracts from readability.

There are several places below that have the same problem.  I realize this is from @pstaabp's pull request.   Sorry he did that.  He should know better.

drgrice1 reviewed (approved):
I see some issues caused by @pstaabp's POD changes.  Once that is fixed,  I think this pull request looks good.

Alex-Jordan commented:
I shortened things. I generally try to end by character 99, with character 100 as the newline. The lorem ipsum text was an exception since the alternatives with perl for a long string are also ugly. But I changed to one such alternative.

drgrice1 commented:
Yeah, that is better.  Thanks.

drgrice1 commented:
It seems there is a problem with the "texpre" option after this pull request.  I have attached a minimal working example.  In this example I want a border on the right of the first cell in the second row, and a line (border) above all but the first cell of that row to make it look like long division.  This works in html, but fails in hardcopy.

The problem is that the TeX output has a multicol on the first cell and puts the cline inside that multicol instead of before it.

I have seen this same problem in some other examples.  In one case the cell has a column header and the `\bfseries` for that was placed before the `texpre` option value.

So in general it seems that the `texpre` is not quite "pre".

[long-division.pg.txt](https://github.com/openwebwork/pg/files/10968988/long-division.pg.txt)

Alex-Jordan commented:
I think you're trying to apply a modifier to a batch of cells (2 through 5) inside the options argument for a single cell. And it's not going to work that way. `texpre` and `texpost` are only meant to affect the cell in question. Some other options affect the row containing the cell in question. But those should be the only paradigms for cell options: affect only that cell or affect the entire row. 

But good news. An improvement from the old niceTables lets you do this, with HTML and LaTeX all in one. I didn't try it, but something like this should work, without getting into CSS:

```
			[
				[ 'x^2 - 9', halign => 'r|'],
				[ '4x^3', top => 1 ],
				[ '+4x^2', top => 1 ],
				[ '-28x', top => 1],
				[ '-30', top => 1 ]
			],
```

If it were not the top row, then you would have to use `bottom` arguments in the previous row. And if neither of these things work, then I'd say there is a bug for me to look into.

Alex-Jordan commented:
Also, I would hope this works:

```
			[
				[ 'x^2 - 9', halign => 'r|', rowtop => 1, top => 0],
				[ '4x^3'],
				[ '+4x^2'],
				[ '-28x'],
				[ '-30']
			],
```

But it could be tricky expecting the cell's 0 value to override the row's 1px value. In LaTeX output especially.

drgrice1 commented:
This is the second row though.  Instead you have to add `rowbottom => 1` to the row above it.  I discovered that after I posted here.  I didn't try your second approach.

Alex-Jordan commented:
Sorry, I wasn't seeing that actual first row when I peeked at the code. So, ideally this would work:
```
	DataTable([
			[ ['', rowbottom => 1, bottom => 0], '', '', '4x', '+4' ],
                        ...
```

But it doesn't. I suspect that `0` is evaluated as false at some point and based on that, nothing is done to write the bottom border as being 0px. I can try to fix that.

In the meantime, you probably already found this, but this works:
```
	DataTable([
                        [
                                '',
                                ['', bottom => 1],
                                ['', bottom => 1],
                                ['4x', bottom => 1],
                                ['{}+4', bottom => 1]
                        ],
			[
				[ 'x^2 - 9', halign => 'r|'],
                                '4x^3', '{}+4x^2', '{}-28x', '{}-30'
			],
			[
				'',
				[ '4x^3', bottom => 1],
				[ '', bottom => 1],
				[ '{}-36x', bottom => 1]
			],
			['', '', '4x^2', '{}+8x', '{}-30'],
			[
				'', '',
			   	['4x^2', bottom => 1],
				['', bottom => 1],
				['{}-36', bottom => 1]
			],
			['', '', '', '8x', '{}+6'],
		], align => 'rrrrr', encase => ['\(', '\)'], booktabs => 0
	)
```

Note that I disabled booktabs. I also put some `{}` in their so the `+` and `-` are treated as addition/subtraction, not positive/negative.

drgrice1 commented:
Yeah, I had it working with `bottom => 1` on the top row.  I also had to add `booktabs => 0` to prevent disjointed line segments.  It is unfortunate that the booktabs creator doesn't like vertical lines in tables, and doesn't fix it to work with that.  This situation is not exactly what the booktabs creator is thinking about though.

I know about adding braces to make positives/negatives become addition/subtraction, but this was just a minimal working example.  In the actual problem, there are also not hard coded numbers.

Alex-Jordan commented:
It seems that it will not be possible (without some tedious work) to use `rowbottom` to put a line under an entire row, and then use an individual cell `bottom => 0` to erase one section. Suppose the CSS for a row puts a bottom border on the row. Experimentally, it seems the CSS for a cell border draws a new separate line. It does not replace a segment from the row's border. So setting a cell to `bottom => 0` just will not erase a section of the row border.

This could only be made to work if the `rowbottom` setting did not actually apply CSS to the HTML row. And instead, applied CSS to each individual cell. And then that CSS could be replaced by an individual cell's `bottom` value.

I'm not sure it is worth the effort. Not only the effort, but for what it's worth, it would clutter the HTML to have lots of CSS for each cell's bottom border when logically one piece of CSS for the row should do. The advice in the POD could be that if you use `rowbottom` or `rowtop`, you are expected to not need a `bottom` or `top` anywhere within the row. Otherwise, just use as many `bottom` or `top` within the row as you need.

drgrice1 commented:
Yeah, I don't think that is a good approach either.  I am good with the first approach.

Alex-Jordan commented:
There is more wrong with the POD following Peter's otherwise helpful edits.
Like the text around line 21 describing LayoutTable looks like it fell into
a copy-paste blender. I will revise it all. Plus I've had a few ideas for
improvements.

On Tue, Mar 14, 2023 at 3:47 PM Glenn Rice ***@***.***> wrote:

> Yeah, I don't think that is a good approach either. I am good with the
> first approach.
>
> —
> Reply to this email directly, view it on GitHub
> <https://protect2.fireeye.com/v1/url?k=31323334-501d2dca-3132feb7-454455534531-d4cd758ed04b4d00&q=1&e=daa7aa7e-0ad7-4438-8ca0-b73020579bc5&u=https%3A%2F%2Fgithub.com%2Fopenwebwork%2Fpg%2Fpull%2F752%23issuecomment-1468969163>,
> or unsubscribe
> <https://protect2.fireeye.com/v1/url?k=31323334-501d2dca-3132feb7-454455534531-ac38e9c6b7e46796&q=1&e=daa7aa7e-0ad7-4438-8ca0-b73020579bc5&u=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FABEDOAAKVQ6TSJAFK5CNWATW4DYRLANCNFSM6AAAAAASZCMM2E>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Alex Jordan
Mathematics Instructor
Portland Community College
